src/main/java/org/csanchez/jenkins/plugins/kubernetes/KubernetesCloud.java

```diff
@@ -619,126 +688,330 @@ private void logLastLines(List<ContainerStatus> containers, String podId, String
-                                // Pod is waiting for some reason
-                                LOGGER.log(Level.INFO, "Container is waiting {0} [{2}]: {1}",
-                                        new Object[] { podId, info.getState().getWaiting(), info.getName() });
-                                // break;
-                            }
-                            if (info.getState().getTerminated() != null) {
-                                terminatedContainers.add(info);
-                            } else if (!info.getReady()) {
-                                allContainersAreReady = false;
-                            }
-                        }
-                    }
+        protected Pod createPod(KubernetesClient client, Pod podTemplate, String namespace) {
+            return client.pods().inNamespace(namespace).create(podTemplate);
+        }
 
-                    if (!terminatedContainers.isEmpty()) {
-                        Map<String, Integer> errors = terminatedContainers.stream().collect(Collectors.toMap(
-                                ContainerStatus::getName, (info) -> info.getState().getTerminated().getExitCode()));
+        protected String getNamespace(KubernetesClient client) {
+            return Strings.isNullOrEmpty(t.getNamespace())
+                    ? client.getNamespace()
+                    : t.getNamespace();
+        }
 
-                        // Print the last lines of failed containers
-                        logLastLines(terminatedContainers, podId, namespace, slave, errors);
-                        throw new IllegalStateException("Containers are terminated with exit codes: " + errors);
-                    }
+        protected String getPodId(Pod pod) {
+            return pod.getMetadata().getName();
+        }
 
-                    if (!allContainersAreReady) {
-                        continue;
-                    }
-                    
-                    if (validStates.contains(pod.getStatus().getPhase())) {
-                        break;
-                    }
-                }
-                String status = pod.getStatus().getPhase();
-                if (!validStates.contains(status)) {
-                    throw new IllegalStateException("Container is not running after " + j + " attempts, status: " + status);
-                }
+        protected String getPodNameSpace(Pod pod) {
+            return pod.getMetadata().getNamespace();
+        }
 
-                j = t.getSlaveConnectTimeout();
+        private SlaveOperationDetails waitForSlaveToComeOnline(String podId, String namespace, String slaveName) throws Exception {
+            Slave slave = (Slave) jenkins().getNode(slaveName);
+            Pod pod = getPodByNamespaceAndPodId(namespace, podId);
+            int slaveConnectTimeoutInSeconds = t.getSlaveConnectTimeout();
+            int attemptsToMake = slaveConnectTimeoutInSeconds / SlaveTimeLimitedTaskRunner.DEFAULT_SLEEP_AFTER_FAILURE_IN_MILLIS;
```

Comment:
That's what you get when you don't have unit tests, haha! I'm going to cover it now